### PR TITLE
feat(mongoose-nested-fields): generate flattened fields in schema

### PR DESCRIPTION
**Pull request by SteveBunlon**
## Pull Request checklist:

- [x] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org)
- [x] Create automatic tests
- [x] Test manually the implemented changes
- [x] Review my own code (indentation, syntax, style, simplicity, readability)
- [x] Wonder if you can improve the existing code
- [x] If needed, Types have been updated

Types have already been updated here --> https://github.com/ForestAdmin/forest-express-mongoose/pull/619/files

### CODE REVIEW

File: test/tests/services/fields-flattener.test.js [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+  describe('validating the flatten property', () => {
+    describe('when the flatten property is not an array', () => {
+      it('should display an error message', () => {
+        expect.assertions(1);
+
+        jest.resetAllMocks();
+        const fieldsFlattener = new FieldsFlattener({}, { });
+        fieldsFlattener.validateOptions();
+
+        expect(errorLoggerSpy).toHaveBeenCalledTimes(1);
```
Comment (olesyakorovina) [inline test/tests/services/fields-flattener.test.js:47]: would be nice to validate the error messages as well in all the logger tests ;)

File: test/tests/services/fields-flattener.test.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+    describe('when the field provided is invalid', () => {
+      describe('when the field does not exist', () => {
+        describe('when the field is of type string', () => {
+          it('should display an warning message', () => {
+            expect.assertions(1);
+
+            jest.resetAllMocks();
+            const fieldsFlattener = new FieldsFlattener({ fields: [] }, ['test']);
+            fieldsFlattener.validateOptions();
```
Comment (olesyakorovina) [inline test/tests/services/fields-flattener.test.js:59]: maybe this code can be refactored in a `beforeAll` in the above `describe`?

File: test/tests/services/fields-flattener.test.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+          });
+        });
+
+        describe('when the field is of type object', () => {
+          it('should display a warning message', () => {
+            expect.assertions(1);
+
+            jest.resetAllMocks();
+            const fieldsFlattener = new FieldsFlattener({ fields: [] }, [{ field: 'test' }]);
+            fieldsFlattener.validateOptions();
```
Comment (olesyakorovina) [inline test/tests/services/fields-flattener.test.js:80]: same here

File: test/tests/services/fields-flattener.test.js [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+  describe('flattening a field', () => {
+    it('should merge fields name with | as separator', () => {
+      expect.assertions(1);
+
+      const schema = generateEngineSchema();
+
+      const fieldsFlattener = new FieldsFlattener(schema, ['engine']);
+      fieldsFlattener.flattenFields();
+
+      expect(schema.fields[0].field).toContain('|');
```
Comment (olesyakorovina) [inline test/tests/services/fields-flattener.test.js:128]: would be nice to validate the whole field name

File: test/tests/services/fields-flattener.test.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+          expect(schema.fields).toHaveLength(5);
+        });
+      });
+      describe('when the level property is lower than actual level', () => {
+        it('should flatten only until specified level', () => {
+          expect.assertions(1);
+
+          const schema = generateEngineSchema();
+          const fieldsFlattener = new FieldsFlattener(schema, [{ field: 'engine', level: 0 }]);
```
Comment (olesyakorovina) [inline test/tests/services/fields-flattener.test.js:164]: maybe rather test with `level: 1` to not repeat the following test?

File: test/tests/services/fields-flattener.test.js [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+        const schema = generateEngineSchema();
+        const fieldsFlattener = new FieldsFlattener(schema, [{ field: 'engine', level: 0 }]);
+
+        fieldsFlattener.flattenFields();
+
+        expect(schema.fields).toHaveLength(4);
+      });
+    });
+    describe('when the level property is under 0', () => {
```
Comment (olesyakorovina) [inline test/tests/services/fields-flattener.test.js:184]: what happens if I pass a string or an object to it? for example `level: '0'` seems to work, but `level: 'test'` ignores completely the flattening. I think we should cover those use-cases with tests so the behavior wouldn't change in the future

**Review by olesyakorovina — CHANGES_REQUESTED** [review]
The implementation works great, just some small comments about the tests

File: test/tests/services/fields-flattener.test.js (hunk 1 above)
Comment (SteveBunlon) [reply test/tests/services/fields-flattener.test.js:47]: 👍

File: test/tests/services/fields-flattener.test.js (hunk 2 above)
Comment (SteveBunlon) [reply test/tests/services/fields-flattener.test.js:59]: 👍

File: test/tests/services/fields-flattener.test.js (hunk 3 above)
Comment (SteveBunlon) [reply test/tests/services/fields-flattener.test.js:80]: 👍

File: test/tests/services/fields-flattener.test.js (hunk 4 above)
Comment (SteveBunlon) [reply test/tests/services/fields-flattener.test.js:128]: 👍

File: test/tests/services/fields-flattener.test.js (hunk 5 above)
Comment (SteveBunlon) [reply test/tests/services/fields-flattener.test.js:164]: Indeed, as a consequence I have updated the data set for the test suit.

File: test/tests/services/fields-flattener.test.js (hunk 6 above)
Comment (SteveBunlon) [reply test/tests/services/fields-flattener.test.js:184]: Changed 👍 Now the level is strongly parsed and if the parsing stuff goes wrong, I remove the level and fallback to infinite flattening.

File: test/tests/services/fields-flattener.test.js [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+      });
+    });
+
+    describe('when the level property is not a number', () => {
+      it('should display a warning message stating that all levels will be flatten', () => {
+        expect.assertions(2);
+
+        jest.resetAllMocks();
+        const fieldsFlattener = new FieldsFlattener(generateEngineSchema(), [{ field: 'engine', level: null }]);
+        fieldsFlattener.validateOptions();
```
Comment (olesyakorovina) [inline test/tests/services/fields-flattener.test.js:140]: Can be factorized I guess

File: test/tests/services/fields-flattener.test.js [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+  });
+
+  describe('flattening a field', () => {
+    it('should merge fields name with | as separator', () => {
+      expect.assertions(1);
+
+      const schema = generateEngineSchema();
+
+      const fieldsFlattener = new FieldsFlattener(schema, ['engine']);
+      fieldsFlattener.flattenFields();
```
Comment (olesyakorovina) [inline test/tests/services/fields-flattener.test.js:166]: Can be factorized

**Review by olesyakorovina — CHANGES_REQUESTED** [review]
I have a crash when I pass a string instead of array to `flatten`

```
[forest] 🌳🌳🌳  Could not flatten fields from collection cars, flatten property should be an array.
{}
[forest] 🌳🌳🌳  An error occured while computing the Forest schema. Your application schema cannot be synchronized with Forest. Your admin panel might not reflect your application models definition. this.flatten.forEach is not a function
{
  "stack": "TypeError: this.flatten.forEach is not a function\n    at FieldsFlattener.flattenFields (/Users/olesyakorovina/Projects/forest-express-mongoose/dist/services/fields-flattener.js:126:20)\n    at Object.exports.collection (/Users/olesyakorovina/Projects/forest-express/dist/index.js:440:23)\n    at Object.<anonymous> (/Users/olesyakorovina/Projects/fa-mongo-cars-v7/dist/forest/cars.js:9:27)\n    at Module._compile (internal/modules/cjs/loader.js:1063:30)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)\n    at Module.load (internal/modules/cjs/loader.js:928:32)\n    at Function.Module._load (internal/modules/cjs/loader.js:769:14)\n    at Module.require (internal/modules/cjs/loader.js:952:19)\n    at require (internal/modules/cjs/helpers.js:88:18)\n    at /Users/olesyakorovina/Projects/forest-express/node_modules/require-all/index.js:56:46\n    at Array.forEach (<anonymous>)\n    at requireAll (/Users/olesyakorovina/Projects/forest-express/node_modules/require-all/index.js:34:9)\n    at loadCollections (/Users/olesyakorovina/Projects/forest-express/dist/index.js:96:3)\n    at _callee$ (/Users/olesyakorovina/Projects/forest-express/dist/index.js:338:15)\n    at tryCatch (/Users/olesyakorovina/Projects/forest-express/node_modules/regenerator-runtime/runtime.js:45:40)\n    at Generator.invoke [as _invoke] (/Users/olesyakorovina/Projects/forest-express/node_modules/regenerator-runtime/runtime.js:274:22)\n    at Generator.prototype.<computed> [as next] (/Users/olesyakorovina/Projects/forest-express/node_modules/regenerator-runtime/runtime.js:97:21)\n    at asyncGeneratorStep (/Users/olesyakorovina/Projects/forest-express/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)\n    at _next (/Users/olesyakorovina/Projects/forest-express/node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)"

**Review by olesyakorovina — APPROVED** [review]

**Review by olesyakorovina — APPROVED** [review]

**Pull request merged**